docs: rewrite Configure Agents index and document external agent harnesses (#2311)
NVIDIA-NeMo/Gym: fern/versions/latest/pages/agent-server/integrate-existing-agents.mdx

---
title: "Integrate Existing Agents"

> description: "Built-in NeMo Gym agents, and the agents that run external harnesses — coding CLIs, agent frameworks, and benchmark scaffolds — against NeMo Gym environments."

Reviewer: description: "Wire built-in NeMo Gym agents servers, and the ones that run external harnesses — coding CLIs, agent frameworks, and benchmark scaffolds — against NeMo Gym environments."

position: 2
---

NeMo Gym ships general-purpose agents that work with any Resources server, and agents that wrap external harnesses so a third-party agent can run against NeMo Gym benchmarks with configuration only. If neither fits, you can keep your agent in your own repo and [drive it remotely](/agent-server/remote-agent).

> ## Built-in NeMo Gym Agents

Reviewer: ```suggestion
## Built-in NeMo Gym Agents Servers
```

> These are native agents with no external dependency. They call the Model server, route tool calls to the Resources server, and verify the result.

Reviewer: ```suggestion
These are native agents servers with no external dependency. They call the Model server, route tool calls to the Resources server, and call the verify() api in the Resource server to score the trajectory.
```

| Agent | Use it when |
|---|---|
| [`simple_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/simple_agent) | You want general-purpose multi-step tool calling with a configurable step budget. Works with any Resources server, and is the usual starting point. |

> | [`non_executing_simple_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/non_executing_simple_agent) | The tool call itself is the answer. Forwards one request to the model and passes the response to the verifier without executing anything. |

Reviewer: | [`non_executing_simple_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/non_executing_simple_agent) | The tool call itself is the answer. Forwards one request to the model and passes the response to the verifier without executing anything. 

That is a weird name and functionality, we should probably review in a future github issue if you want to create it

> | [`tool_simulation_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/tool_simulation_agent) | Tool results should be simulated rather than really executed. |

Reviewer: Also very weird distinction. We should revisit and understand why this is implemented. Please fire a future github issue

> | [`gymnasium_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/gymnasium_agent) | Your environment is a `GymnasiumServer` Resources server and the agent should drive its reset/step loop. |

Reviewer: ```suggestion
| [`gymnasium_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/gymnasium_agent) | Your environment is a `GymnasiumServer` Resources server and you need a compatible Agent Server to drive its reset/step loop. |
```

## External Agent Harnesses

Each of these runs a third-party harness as a NeMo Gym agent server. The harness keeps its own loop, prompts, and tools; NeMo Gym seeds the session, translates the resulting trajectory into Responses API format, and verifies it against the Resources server.

> Because verification stays on the NeMo Gym side, a harness that brings its own tools can often run an existing benchmark by adding a `<benchmark>_<agent>` config, leaving the dataset and verifier unchanged — only the Resources server's task data and `/verify` are used. This is how `hermes_agent` is used with the math, coding, reasoning_gym, MCQA, and instruction-following benchmarks. It does not hold for every harness: agents such as `swe_agents`, `osworld_agent`, and `tau2` are tied to specific benchmarks.

Reviewer: ```suggestion
Because verification stays on the NeMo Gym side, an Agent harness that brings its own tools can often run against an existing benchmark by adding a `<benchmark>_<agent>` config, leaving the dataset and verifier unchanged — only the Resources server's task data and `/verify` are used. This is how `hermes_agent` is used with the math, coding, reasoning_gym, MCQA, and instruction-following benchmarks. It does not hold for every harness: agents such as `swe_agents`, `osworld_agent`, and `tau2` are tied to specific benchmarks.
```

| [`kilocode_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/kilocode_agent) | Kilo Code (`kilo run`) | A fork of OpenCode; mirrors `opencode_agent` but parses the CLI's JSON event stream. |
| [`openclaw_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/openclaw_agent) | OpenClaw (`openclaw agent --local --json`) | |
| [`pi_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/pi_agent) | pi (`pi --print --mode json`) | |

<Note>
Every agent in this table is evaluation-only today: token IDs and log probabilities are not threaded through, so the rollouts are not trainable. All of them except `kilocode_agent` accept an optional `model_server` ref that routes generation through NeMo Gym, which makes the model calls capturable and lets one config run against different backends. `kilocode_agent` takes a provider base URL directly. See [Evaluation-only vs training-grade](#evaluation-only-vs-training-grade).
</Note>

> ### Agent Frameworks

Reviewer: Maybe a better name for this section is External Frameworks, because I see both several Environment Frameworks

Python agent frameworks embedded in the agent server process.

> | Agent | Framework | Notes |

Reviewer: Server

| [`osworld_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/osworld_agent) | OSWorld `DesktopEnv`, with the desktop container/VM lifecycle in the NeMo Gym Docker Sandbox | OSWorld desktop-computer tasks |
| [`toolsandbox_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/toolsandbox_agent) | [ToolSandbox](https://github.com/apple/ToolSandbox) (Apple) | Multi-turn tool use with a user simulator — a natural-language reply continues the conversation instead of ending the rollout. |
| [`tau2`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/tau2) | τ²-bench, checked out and installed at setup | τ²-bench, plus τ³ banking-knowledge configs that reuse the same bridge |
| [`pinchbench`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/pinchbench) | [PinchBench](https://github.com/pinchbench/skill) over [OpenClaw](https://github.com/openclaw/openclaw) | 147 real-world tasks (calendar, email triage, log analysis, coding, research, writing) |

### Agent-Agnostic Task Runners

> These are not harnesses. They take another NeMo Gym agent, run it inside a benchmark's task container, and grade what it leaves behind.

Reviewer: ```suggestion
These are not harnesses. They wire together another NeMo Gym agent server, run it inside a benchmark's task container, and grade what it leaves behind.
```

> | Agent | What it does |

Reviewer: Server

| [`anyterminal_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/anyterminal_agent) | Runs your agent inside a Terminal-Bench task container, then runs the task's `tests/test.sh` in that same container. |

Alongside these, `responses_api_agents/` also holds benchmark-specific custom agents (for example SciCode's per-sub-step loop and CritPt's two-turn solve-then-fill-template flow). Browse the [directory](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents) for the full set; each has a README with its own quickstart.

## How an External Harness Is Wrapped

The wrappers vary in size — the container-based ones do considerably more setup than the CLI ones — but they share four responsibilities. The `swe_agents` opencode integration states the underlying rule as "swap the transport, keep the loop": leave the harness's prompts, tools, and control flow alone, and change only where it gets tokens from and where its trajectory goes.

> 1. Run the harness, inside the agent server's `run()`. A subprocess for a CLI, an in-process import for a Python framework, a container for a sandboxed benchmark.

Reviewer: I think this might not be totally true. I think you need to be more precise. the run() is responsible for orchestrating part of the Resource Server seed session and the verifier. potentially the harness is part of the "responses()"

| [`gymnasium_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/gymnasium_agent) | Drives the reset/step loop of a `GymnasiumServer` Resources server. |

## Agent Servers That Run a Third-Party Harness

In these, the external harness owns the agent loop and keeps its own prompts and tools. NeMo Gym usually seeds the session, translates the resulting trajectory into Responses API format, and verifies it against the Resources server; a few whose benchmark harness owns scoring compute the reward themselves.

Because verification stays on the NeMo Gym side, an agent harness that brings its own tools can often run in an existing environment by adding a `<benchmark>_<agent>` config, leaving the dataset and verifier unchanged — only the Resources server's task data and `/verify` are used. This is how `hermes_agent` is used with the math, coding, reasoning_gym, MCQA, and instruction-following benchmarks. It does not hold for every harness: agent servers such as `swe_agents`, `osworld_agent`, and `tau2` are tied to specific benchmarks.

> ### Coding CLIs

Reviewer 2: Coding CLIs is a bit misleading. A few other options to consider, but we should ensure we are consistent -- and using CLI could also be confusing. Suggestions:

1. External Agent Harness — emphasizes they're third-party, but "External" is already covered by the parent section heading. 
2. External Harnesses — most accurate: all six run their own agentic loop as a subprocess. Captures both the CLI and agentic nature without implying "coding."

Agent servers that run a coding CLI as a subprocess. The CLI executes its own tools, and its output is parsed back into NeMo Gym format.

> | Agent Server | Harness | Notes |

Reviewer 2: Nit: can we rename `Notes` to `Description` or `Summary`?

| [`codex_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/codex_agent) | OpenAI Codex CLI (`codex exec`) | Runs against OpenAI, or against any NeMo Gym Model server, which serves the streaming Responses dialect Codex speaks. Also supports Agent Skills. |
| [`opencode_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/opencode_agent) | OpenCode (`opencode run`) | Reads the run's SQLite session to recover the full trajectory including tool calls. |
| [`kilocode_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/kilocode_agent) | Kilo Code (`kilo run`) | A fork of OpenCode; mirrors `opencode_agent` but parses the CLI's JSON event stream. |
| [`openclaw_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/openclaw_agent) | OpenClaw (`openclaw agent --local --json`) | |

> | [`pi_agent`](https://github.com/NVIDIA-NeMo/Gym/tree/main/responses_api_agents/pi_agent) | pi (`pi --print --mode json --no-session`) | |

Reviewer 2: Missing notes/description
